lib/resources/aws/aws_ec2_instances.rb

```diff
@@ -0,0 +1,72 @@
+# author: Christoph Hartmann
+class AwsEc2Instances < Inspec.resource(1)
+  name 'aws_ec2_instances'
+  desc 'Verifies settings for an EC2 instance'
+  supports platform: 'aws'
+  example "
+    describe_ec2_instances.where { public_ip_address !=nil } do
+      its('instance_id') { should eq [] }
+    end
+  "
+
+  # TODO: rewrite to avoid direct injection, using AwsPluralResourceMixin
+  def initialize(opts = {}, conn = nil)
+    @opts = opts
+    @ec2_client = conn ? conn.ec2_client : inspec_runner.backend.aws_client(Aws::EC2::Client)
+  end
+
+  # Set up some accessors and the filter table
+  filter = FilterTable.create
+  filter.add_accessor(:where).add_accessor(:entries)
+  Aws::EC2::Types::Instance.members.each do |field|
+    # Add a filter table field for each AWS instance property
+    filter.add(field, field: field)
+    # Add an accessors to access properties without the filter table
+    define_method(field) do
+      list.map(&:field)
+    end
+  end
+  filter.connect(self, :list)
+
+  # Give uses of `inspec shell` a few hints about what they can test
+  def possible_properties
+    Aws::EC2::Types::Instance.members
+  end
+
+  def inspec_runner
+    inspec if respond_to?(:inspec)
+  end
+
+  def to_s
+    'EC2 Instances'
+  end
+
+  # Calls self.class.list to gain access to 'class instance variables'
+  def list
+    self.class.list(@ec2_client)
+  end
+
+  def self.list(ec2_client)
+    # Make sure we only call depaginate_api once
+    @aws_ec2_instances_cache ||= depaginate_api(ec2_client)
```

Comment:
I think this violates our caching philiosophy.  To date, I think we've been careful to cache only connections, and avoid caching any data returned from an API.  We don't know how long an InSpec run may live - if they are scanning S3 on a per-object basis, it could be hours - and this would result in stale EC2 data.  Additionally, to date we've treated Train as being the component responsible for providing a cache authority (with facilities for disabling, expiring, etc).  @jquick, @chris-rock , what do you think?

Reply:
We should probably review that policy...

Take a look at performance of `aws_iam_users`. 

- It takes approx 1 min to pull all the users in an AWS account with 215 users.
- If I use `describe aws_iam_users` twice, it runs for twice the time.
- If I modify it to cache at class level instead of instance level its runtime remains roughly constant.

Lazy loading in the filter table will help, but it still takes significant time for the API calls and generates a ton of network traffic. Not sure we really care about data being a few seconds/minutes stale.

Reply:
We had a chat in the InSpec team standup on 2018-03-19, and my concerns were mostly assuaged.  

* You can disable Train caching with a CLI option to `inspec exec`
* We can add a doc warning that results may be cached
* Relatively few users would fetch EC2 instances, do something long running, then check EC2 instances again (I'd love to hear a common use case if you have one)
* With caching enabled, we can think of the scan as being a point-in-time compliance check.  Without caching, it's like multiple scans.
* It is possible to clear the cache, and we can document that better so that affected users will have a workaround.

All that said, my only concern is where the code is located - I think we want caching in the general case in Train.  @jquick , any advice?

Reply:
Refs #3127